nat/traversal/pinger.go

```diff
@@ -0,0 +1,207 @@
+	localPort     int
+	pingTarget    chan json.RawMessage
+	pingReceived  chan bool
+	pingCancelled chan bool
+	eventsTracker *EventsTracker
+	configParser  ConfigParser
+}
+
+// NewPingerFactory returns Pinger instance
+func NewPingerFactory(tracker *EventsTracker, parser ConfigParser) *Pinger {
+	target := make(chan json.RawMessage)
+	received := make(chan bool)
+	cancel := make(chan bool)
+	return &Pinger{
+		pingTarget:    target,
+		pingReceived:  received,
+		pingCancelled: cancel,
+		eventsTracker: tracker,
+		configParser:  parser,
+	}
+}
+
+// Start starts NAT pinger and waits for pingTarget to ping
+func (p *Pinger) Start() {
+	log.Info(prefix, "Starting a NAT pinger")
+	for {
+		select {
+		case dst := <-p.pingTarget:
+			IP, port, err := p.configParser.Parse(dst)
+			log.Infof("%s ping target received: IP: %v, port: %v", prefix, IP, port)
+			if port == 0 {
+				// client did not sent its port to ping to, attempting with service start
+				p.pingReceived <- true
+				continue
+			}
+			conn, err := p.getConnection(IP, port)
+			if err != nil {
+				log.Error(errors.Wrap(err, "failed to get connection"))
+				continue
+			}
+			go p.ping(conn)
+			time.Sleep(pingInterval * time.Millisecond)
+			p.pingReceiver(conn)
+			p.pingReceived <- true
+			log.Info(prefix, "ping received, waiting for a new connection")
+			conn.Close()
+		}
+	}
+
+	return
+}
+
+// PingProvider pings provider determined by destination provided in sessionConfig
+func (p *Pinger) PingProvider(ip string, port int) error {
+	log.Info(prefix, "NAT pinging to provider")
+
+	conn, err := p.getConnection(ip, port)
+	if err != nil {
+		return errors.Wrap(err, "failed to get connection")
+	}
+
+	//let provider ping first
+	time.Sleep(20 * time.Millisecond)
+	go p.ping(conn)
+	time.Sleep(pingInterval * time.Millisecond)
+	p.pingReceiver(conn)
+	// wait for provider to startup
+	time.Sleep(3 * time.Second)
+	conn.Close()
+
+	return nil
+}
+
+func (p *Pinger) ping(conn *net.UDPConn) error {
+	n := 1
+
+	for {
+		select {
+		case <-p.pingCancelled:
+			return nil
+
+		case <-time.After(pingInterval * time.Millisecond):
+			log.Info(prefix, "pinging.. ")
+			if n > 4 {
+				n = 128
+			}
+
+			ipv4.NewConn(conn).SetTTL(n)
+			n++
+
+			_, err := conn.Write([]byte("continuously pinging to " + conn.RemoteAddr().String()))
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (p *Pinger) getConnection(ip string, port int) (*net.UDPConn, error) {
+	udpAddr, err := net.ResolveUDPAddr("udp4", fmt.Sprintf("%s:%d", ip, port))
+	if err != nil {
+		return nil, err
+	}
+
+	log.Info(prefix, "remote socket: ", udpAddr.String())
+
+	conn, err := net.DialUDP("udp", &net.UDPAddr{Port: p.localPort}, udpAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Info(prefix, "local socket: ", conn.LocalAddr().String())
+
+	return conn, nil
+}
+
+// PingTargetChan returns a channel which relays ping target address data
+func (p *Pinger) PingTargetChan() chan json.RawMessage {
```

Comment:
don't really need to return the channel here, could instead accept a ping target and push it to the channel ourselves.